File: dynamicProgramming/medium/983. Minimum Cost For Tickets.py
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        set_days = set(days)
        d = dict()
        c = {0:1, 1:7, 2:30}
        before_day = 0
        d[0] = 0

        for day in range(1, days[-1]+1):
            if day in set_days:
                result = float('inf')
                for j in range(len(costs)):
                    cost = costs[j]
                    interval = c[j]
                    if day >= interval:
                        result = min(result, d[day-interval]+cost)
                    else:
                        result = min(result, cost)
                d[day] = result
                before_day = day
            else:
                d[day] = d.get(before_day)

        return d.get(days[-1])

File: dynamicProgramming/medium/983. Minimum Cost For Tickets.java
class Solution {
    public int mincostTickets(int[] days, int[] costs) {
        int lastDay = days[days.length-1];

        boolean[] travel = new boolean[lastDay + 1];
        for (int day : days) {
            travel[day] = true;
        }

        int[] dp = new int[lastDay + 1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = 0;

        for (int i = 1; i <= lastDay; i++) {
            if (!travel[i]) {
                dp[i] = dp[i-1];
                continue;
            }
            dp[i] = Math.min(dp[i], dp[i-1] + costs[0]);
            dp[i] = Math.min(dp[i], (i >= 7 ? dp[i-7] : 0) + costs[1]);
            dp[i] = Math.min(dp[i], (i >= 30 ? dp[i-30] : 0)  + costs[2]);
        }

        return dp[lastDay];
    }
}